collector/netframework_clr.go

```diff
@@ -0,0 +1,1271 @@
+		),
+		StackWalkDepth: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, securitySubsystem, "stack_walk_depth"),
+			"Displays the depth of the stack during that last runtime code access security check.",
+			[]string{"process"},
+			nil,
+		),
+		TotalRuntimeChecks: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, securitySubsystem, "runtime_checks_total"),
+			"Displays the total number of runtime code access security checks performed since the application started.",
+			[]string{"process"},
+			nil,
+		),
+
+		// Process name whitelist and blacklist
+		processWhitelistPattern: regexp.MustCompile(fmt.Sprintf("^(?:%s)$", *netclrWhitelist)),
+		processBlacklistPattern: regexp.MustCompile(fmt.Sprintf("^(?:%s)$", *netclrBlacklist)),
+	}
+
+	netclrCollector.netclrCollectors = netclrCollector.getNETCLRCollectors()
+	if *netclrPrintCollectors {
+		fmt.Printf("Available .NET CLR sub-collectors:\n")
+		for name := range netclrCollector.netclrCollectors {
+			fmt.Printf(" - %s\n", name)
+		}
+		os.Exit(0)
+	}
+
+	return netclrCollector, nil
+}
+
+func (c *netclrCollector) execute(ctx *ScrapeContext, name string, fn netclrCollectorFunc, ch chan<- prometheus.Metric, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	begin := time.Now()
+	_, err := fn(ctx, ch)
+	duration := time.Since(begin)
+	var success float64
+
+	if err != nil {
+		log.Errorf("netframework_clr sub-collector %s failed after %fs: %s", name, duration.Seconds(), err)
+		success = 0
+		c.netclrChildCollectorFailure++
+	} else {
+		log.Debugf("netframework_clr sub-collector %s succeeded after %fs.", name, duration.Seconds())
+		success = 1
+	}
+	ch <- prometheus.MustNewConstMetric(
+		c.netclrScrapeDurationDesc,
+		prometheus.GaugeValue,
+		duration.Seconds(),
+		name,
+	)
+	ch <- prometheus.MustNewConstMetric(
+		c.netclrScrapeSuccessDesc,
+		prometheus.GaugeValue,
+		success,
+		name,
+	)
+}
+
+func (c *netclrCollector) Collect(ctx *ScrapeContext, ch chan<- prometheus.Metric) error {
+	wg := sync.WaitGroup{}
+
+	c.netclrChildCollectorFailure = 0
+	enabled := netclrExpandEnabledCollectors(*netclrEnabledCollectors)
+	for _, name := range enabled {
+		function := c.netclrCollectors[name]
+
+		wg.Add(1)
+		go c.execute(ctx, name, function, ch, &wg)
+	}
+	wg.Wait()
+
+	// This should return an error if any collector encountered an error.
+	if c.netclrChildCollectorFailure > 0 {
+		return errors.New("at least one child collector failed")
+	}
+	return nil
+}
+
+func (c *netclrCollector) netclrMapProcessName(name string, nameCounts map[string]int) (string, bool) {
+	// Skip the _Global_ instance since it's just a sum of the other instances.
+	if name == "_Global_" {
+		return name, false
+	}
+
+	// Append a "#1", "#2", etc., suffix if a process name appears more than once.
+	// WMI does this automatically, but Perflib does not.
+	procnum, exists := nameCounts[name]
+	if exists {
+		nameCounts[name]++
+		name = fmt.Sprintf("%s#%d", name, procnum)
+	} else {
+		nameCounts[name] = 1
+	}
```

Comment:
I did some testing on this after the discussions in the other PR - on my machine at least, I do not get a stable naming with this approach. The data returned by perflib appears to come sorted by PID, maybe? Not sure this is the ordering, but it is at least not by start order. So, starting a new process with the same name as existing ones may "displace" them in the naming, so `#2` might be `#3`, and so on. And a process terminating will also cause re-ordering, of course.

This also appears to happen with WMI, so it is not a problem introduced by perflib as such, but we should probably try to fix it if we can. From what I can tell, the performance object for the .NET CLR stuff has a `Process ID` field, so we do have that data to use. Two possible paths:
- Add a `pid` label
- Use the process id as the #-suffix

Any reasons not to use the first option?